src/main/java/org/killbill/billing/plugin/adyen/api/mapping/UserDataMappingService.java

```diff
@@ -0,0 +1,196 @@
+        userData.setEmail(email);
+
+        // determine first Name
+        final String propertyFirstName = PluginProperties.findPluginPropertyValue(PROPERTY_FIRST_NAME, properties);
+        final Optional<String> optionalFirstName = toFirstName(propertyFirstName, account);
+        final String firstName = optionalFirstName.isPresent() ? optionalFirstName.get() : null;
+        userData.setFirstName(firstName);
+
+        // determine last Name
+        final String propertyLastName = PluginProperties.findPluginPropertyValue(PROPERTY_LAST_NAME, properties);
+        final Optional<String> optionalLastName = toLastName(propertyLastName, account);
+        final String lastName = optionalLastName.isPresent() ? optionalLastName.get() : null;
+        userData.setLastName(lastName);
+
+        // set ip
+        userData.setIP(PluginProperties.findPluginPropertyValue(PROPERTY_IP, properties));
+
+        return userData;
+    }
+
+    /**
+     *  This function determines the customer id that will be used in the communication with Adyen.
+     *
+     *  The following heuristic is used to determine it:
+     *  1. If there was a customerId in the plugin properties this will be used.
+     *  2. If the account has an externalKey this will be used.
+     *  3. The id of the account is used.
+     *  4. In the unlikely case that there is no account, this function returns an empty Optional.
+     *
+     * @param customerId the customerId that has been sent in the plugin properties
+     * @param account the Kill Bill account
+     * @return the customer id as an Optional
+     */
+    public static Optional<String> toCustomerId( String customerId, Account account) {
+        if (customerId != null) {
+            return Optional.of(customerId);
+        } else if (account != null) {
+            if (account.getExternalKey() != null) {
+                return Optional.of(account.getExternalKey());
+            } else {
+                return Optional.of(account.getId().toString());
+            }
+        } else {
+            return Optional.absent();
+        }
+    }
+
+    /**
+     * This function determines the customer locale that will be used in the communication with Adyen.
+     *
+     * The following heuristic is used to determine it:
+     * 1. if there is a local provided in the plugin properties this will be used
+     * 2. if there the account has a locale, that will be used
+     * 3. if there should be no account or an account without locale, an empty optional will be returned
+     *
+     * @param propertyLocaleString the locale that has been sent in the plugin properties
+     * @param account the Kill Bill account
+     * @return the locale as an Optional
+     */
+    public static Optional<Locale> toCustomerLocale(String propertyLocaleString, Account account) {
+        if (propertyLocaleString != null) {
+            return Optional.of(Locale.forLanguageTag(propertyLocaleString));
```

Comment:
It is also strange from my point of view. I checked a little bit in more detail and found out that the getCountry of the locale would always return null. For both, for forLanguageTag and for new Locale(account.getLocale). I replaced it with LocalUtils from Apache Commons. And it also works with Java 6.